### PR TITLE
Pass {fit,validate,test,predict} to setup() and teardown()

### LINKED ISSUE #5658: Pass all running stages to DataModule.setup
## 🚀 Feature

Currently. `DataModule.setup` is only called with stages `fit` or `test`. But we have several more:

Stages:

https://github.com/PyTorchLightning/pytorch-lightning/blob/5f3372871a333c3229968f1af1b10a925d7ec3ec/pytorch_lightning/trainer/states.py#L39-L49

Note that it's a bit tricky because `fit` is not a `RunningStage`. It indicates `train` or `eval`

### Motivation

Allows having custom logic for each stage

### Pitch

```python
def setup(stage: Optional[str] = None):
    assert stage in list(RunningStage)
    ...
```

### Additional context

We are passing 'test' when predicting as seen in #5579
https://github.com/PyTorchLightning/pytorch-lightning/blob/9137b16068fe03e6db8df548235363e5f5476aac/pytorch_lightning/trainer/trainer.py#L909

### LINKED ISSUE #6376: When I called Trainer.test, some hooks fit are also called
## 🐛 Bug

When I called Trainer.test, some hooks fit are also called, which including: on_fit_start, on_fit_end, and teardown(stage=fit)
I found a similar issue of setup(stage=fit), that bug has been fixed, but seems these hooks also have this bug
https://github.com/PyTorchLightning/pytorch-lightning/issues/2715#issue-665835081

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1GA2Uo8cBaziqoBUI5K2r5x5BjjaZQRJe?usp=sharing

### Expected behavior

during trainer.test, on_fit_start, on_fit_end, teardown(stage=fit) should not be called
（also could you please check if other hooks have this bug?)

### Environment

* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.7.1+cu101
	- pytorch-lightning: 1.2.2
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

- Pass `{'fit','validate','test','predict'}` to `setup(stage=...)` and `teardown(stage=...)`
- Refactor teardown logic
- Fix `.on_fit_{start,end}()` getting called during `trainer.test`
- Fix `.teardown(stage='fit')` getting called during `trainer.test`
- Annotate `stage` argument for `.setup()` and `.teardown()` as `Optional`
- If the `Trainer` is interrupted, `.teardown(stage=None)` will be called

Fixes #5658, fixes #6376

Last step before `trainer.validate()` #4948 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -880,8 +878,6 @@ def test(
             self.__evaluate_using_weights(model, ckpt_path=ckpt_path, dataloaders=test_dataloaders)
         )
 
-        self.teardown('test')
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:883]: why is this removed?

File: pytorch_lightning/trainer/model_hooks.py [hunk 2]
```diff
@@ -22,12 +22,6 @@ class TrainerModelHooksMixin(ABC):
 
     lightning_module: LightningModule
 
-    def is_function_implemented(self, f_name, model=None):
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/model_hooks.py:25]: Okay. I can add it back. No longer used in our codebase

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:883]: No longer needed. This is duplicated

File: pytorch_lightning/core/hooks.py [hunk 3]
```diff
@@ -25,14 +25,14 @@
...
+    def setup(self, stage: Optional[str] = None) -> None:
         """
-        Called at the beginning of fit and test.
+        Called at the beginning of fit (train + validate), validate, test, predict, or tune.
         This is a good hook when you need to build models dynamically or adjust something about them.
         This hook is called on every process when using DDP.
 
         Args:
-            stage: either 'fit' or 'test'
+            stage: either ``'fit'``, ``'validate'``, ``'test'``, or ``'predict'``
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/core/hooks.py:35]: I wasn't sure of passing `str` or the `Enum`. It should be compatible but would you then update all the signatures to indicate that an Enum is passed? or keep the signature as `Optional[str]`?

File: pytorch_lightning/core/hooks.py [hunk 4]
```diff
@@ -53,12 +53,12 @@ def setup(stage):
...
-    def teardown(self, stage: str) -> None:
+    def teardown(self, stage: Optional[str] = None) -> None:
         """
-        Called at the end of fit and test.
+        Called at the end of fit (train + validate), validate, test, predict, or tune.
 
         Args:
-            stage: either 'fit' or 'test'
+            stage: either ``'fit'``, ``'validate'``, ``'test'``, or ``'predict'``
         """
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/core/hooks.py:62]: Handled here: https://github.com/PyTorchLightning/pytorch-lightning/blob/0a30abf931ec5a5f1127bdf98514df6f489cb735/pytorch_lightning/trainer/trainer.py#L1026-L1047

File: pytorch_lightning/core/hooks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/core/hooks.py:35]: Actually, it should be okay as `isinstance(LightningEnum, str) == True`

File: pytorch_lightning/trainer/model_hooks.py [hunk 5]
```diff
@@ -22,13 +23,14 @@ class TrainerModelHooksMixin(ABC):
...
+    def is_function_implemented(self, f_name: str, model: Optional[LightningModule] = None) -> bool:
+        # note: currently unused - kept as it is public
```
Comment (awaelchli) [inline pytorch_lightning/trainer/model_hooks.py:27]: if it's unused we need to add pragma no cover to exclude it from coverage
or test it anyway xD

File: tests/models/test_hooks.py [hunk 6]
```diff
@@ -483,9 +482,6 @@ def teardown(self, stage: str):
...
         'on_test_end',
         'on_test_model_train',
-        'on_fit_end',
-        'teardown',  # for 'fit'
-        'teardown',  # for 'test'
+        'teardown',
     ]
-
-    assert model2.called == expected
+    assert model.called == expected
```
Comment (awaelchli) [inline tests/models/test_hooks.py:487]: reminder: we should rewrite this test with a model mock for clarity and sanity :))

File: tests/helpers/boring_model.py [hunk 7]
```diff
@@ -151,9 +151,11 @@ def prepare_data(self):
     def setup(self, stage: Optional[str] = None):
         if stage == "fit" or stage is None:
             self.random_train = Subset(self.random_full, indices=range(64))
-            self.random_val = Subset(self.random_full, indices=range(64, 128))
             self.dims = self.random_train[0].shape
 
+        if stage in ("fit", "validate") or stage is None:
+            self.random_val = Subset(self.random_full, indices=range(64, 128))
+
```
Comment (awaelchli) [inline tests/helpers/boring_model.py:158]: note this makes the boring model less boring!

**Review by awaelchli — APPROVED** [review]

File: tests/models/test_hooks.py (hunk 6 above)
Comment (awaelchli) [reply tests/models/test_hooks.py:487]: want me to work on it?

File: docs/source/extensions/datamodules.rst [hunk 8]
```diff
@@ -80,7 +80,7 @@ The equivalent DataModule just organizes the same exact code, but makes it reusa
             self.data_dir = data_dir
             self.batch_size = batch_size
 
-        def setup(self, stage=None):
+        def setup(self, stage: Optional[str] = None):
```
[the remark this answers is not in the document]
Comment (carmocca) [reply docs/source/extensions/datamodules.rst:83]: Well we want to improve typing support throughout the codebase, so doing it little by little in each PR feels much better to me.

In this particular case I added it to the docs because I feel it helps a lot when you are reading docs to know what's the type

File: CHANGELOG.md [hunk 9]
```diff
@@ -38,6 +38,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Changed `trainer.evaluating` to return `True` if validating or testing ([#4945](https://github.com/PyTorchLightning/pytorch-lightning/pull/4945))
 
 
+- Changed `setup()` and `teardown()` stage argument to take any of `{fit,validate,test,predict}` ([#6386](https://github.com/PyTorchLightning/pytorch-lightning/pull/6386))
```
[the remark this answers is not in the document]
Comment (carmocca) [reply CHANGELOG.md:41]: Sure

File: pytorch_lightning/trainer/callback_hook.py [hunk 10]
```diff
@@ -29,18 +29,18 @@ class TrainerCallbackHookMixin(ABC):
     callbacks: List[Callback] = []
     lightning_module: LightningModule
 
-    def on_before_accelerator_backend_setup(self, model):
-        """Called in the beginning of fit and test"""
+    def on_before_accelerator_backend_setup(self, model: LightningModule) -> None:
+        """Called at the beginning of fit (train + validate), validate, test, or predict, or tune."""
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/callback_hook.py:33]: https://www.grammarly.com/blog/what-is-the-oxford-comma-and-why-do-people-care-so-much-about-it/

File: pytorch_lightning/trainer/model_hooks.py (hunk 5 above)
Comment (carmocca) [reply pytorch_lightning/trainer/model_hooks.py:27]: I initially removed it: https://github.com/PyTorchLightning/pytorch-lightning/pull/6386#discussion_r588990207

What do you think is the best option?

cc: @Borda

File: tests/models/test_hooks.py (hunk 6 above)
Comment (carmocca) [reply tests/models/test_hooks.py:487]: sure!

File: tests/helpers/boring_model.py (hunk 7 above)
Comment (carmocca) [reply tests/helpers/boring_model.py:158]: A tiny tiny bit :D

File: pytorch_lightning/trainer/callback_hook.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/trainer/callback_hook.py:33]: I didn't know it's called oxford comma
But I guess my question was more why do validate and test not have a "or" in front of it, why is it only the last two elements?

File: pytorch_lightning/core/datamodule.py [hunk 11]
```diff
@@ -77,15 +77,15 @@ def wrapped_fn(*args, **kwargs):
...
+            # If not provided, set call status of 'fit', 'validate', and 'test' to True.
             # We do this so __attach_datamodule in trainer.py doesn't mistakenly call setup('test') on trainer.test()
             stage = args[1] if len(args) > 1 else kwargs.get("stage", None)
 
-            if stage == "fit" or stage is None:
-                obj._has_setup_fit = True
-
-            if stage == "test" or stage is None:
-                obj._has_setup_test = True
+            if stage is None:
```
Comment (ananthsub) [inline pytorch_lightning/core/datamodule.py:84]: - does the wrapper class exist purely to set these attributes and to ensure `prepare_data` is called only on rank 0? 
- does it make sense to add per-stage `setup` and `teardown` methods explicitly on the data module interface? similar to train/val/test/predict dataloaders? or does this become too many hooks? if we have these explicit hooks, then we wouldn't need the `_has_setup_{stage}` attributes right?
- should `stage` be backed by an enum?

File: pytorch_lightning/trainer/callback_hook.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/callback_hook.py:33]: Well, not a native English speaker but would you say

`You can choose soda or beer or water or wine or champagne`

or

`You can choose soda, beer, water, wine, or champagne`

The later seems more natural to me

**Review by ananthsub — COMMENTED** [review]
great fix @carmocca ! everything that has a `setup` should have a `teardown`

File: pytorch_lightning/core/datamodule.py (hunk 11 above)
Comment (carmocca) [reply pytorch_lightning/core/datamodule.py:84]: > does the wrapper class exist purely to set these attributes and to ensure prepare_data is called only on rank 0?

I think so!

> if we have these explicit hooks, then we wouldn't need the _has_setup_{stage} attributes right? and then we could also do away with the wrapper class?

Yes, we wouldn't. setup/teardown are the only hooks that are different in that they take the stage as an argument, which is inconsistent with every other hook

> does it make sense to add per-stage setup and teardown methods explicitly on the data module interface, similar to train/val/test/predict dataloaders? or does this become too many hooks?

It does make sense for consistency, but at this point, deprecating and changing everything does not seem worth it if there is no obvious benefit. Feel free to open a discussion about this for the larger team

Comment (ananthsub) [reply pytorch_lightning/core/datamodule.py:84]: Started a discussion here: https://github.com/PyTorchLightning/pytorch-lightning/discussions/6401

It definitely doesn't block this PR though!

**Review by ananthsub — APPROVED** [review]

**Review by williamFalcon — APPROVED** [review]
